core/objects.go

```diff
@@ -224,6 +224,22 @@ func (r *Registration) MergeUpdate(input Registration) {
 	}
 }
 
+// ValidationTarget describes a single URL and/or IP addresses resolved for the
+// URL/hostname
+type ValidationTarget struct {
+	URL               string   `json:"url,omitempty"`
+	Hostname          string   `json:"hostname,omitempty"`
+	AddressesResolved []net.IP `json:"addressesResolved"`
+	AddressUsed       net.IP   `json:"addressUsed"`
+}
+
+// ValidationTargets describes all of the URLs and/or IP addresses resolved
+// and used during a single validation attempt
+type ValidationTargets struct {
+	Type string             `json:"type"`
```

Comment:
I forgot to explain the reasoning here: It's nice to have a separate struct for each type of ValidationRecord, because otherwise we wind up with a union of fields, of which some only make sense depending on which type of validation we're doing. That makes it hard to write correct code and verify it.